### PR TITLE
Add Distributed PECOS XLinear Modules

### PR DESCRIPTION
**Pull request by weiliw-amz**
*Issue #, if available:*
N/A

*Description of changes:*
Added modules to enable distributed PECOS XLinear training, as well as related unittests.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pecos/distributed/comm/dummy_comm.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#
+#  http://aws.amazon.com/apache2.0/
+#
+#  or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+#  OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+#  and limitations under the License.
+from pecos.distributed.comm.abs_dist_comm import DistComm
+
+
+class DummyComm(DistComm):
```
Comment (jiong-zhang) [inline pecos/distributed/comm/dummy_comm.py]: If we only need this for testing, should this be put in the test folder?

File: pecos/distributed/xmc/xlinear/train.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+def do_train(args):
+    """Distributed CPU training and saving XLinear model"""
+
+    # Distributed communicator
+    mpi_comm = MPIComm()
+
+    # Parse args
+    args_dict = vars(args)
+    train_params = DistributedCPUXLinearModel.TrainParams.from_dict(args_dict, recursive=True)
```
Comment (jiong-zhang) [inline pecos/distributed/xmc/xlinear/train.py:174]: To be consistent with the single node `pecos.xmc.xlinear.train`, we should support input a json file for params.

File: pecos/distributed/comm/dummy_comm.py (hunk 1 above)
Comment (weiliw-amz) [reply pecos/distributed/comm/dummy_comm.py]: The test folder is not a Python module and therefore disable importing. Either put it here for code re-usage, or duplicate this class across `dist_pecos` test modules.

File: pecos/distributed/xmc/xlinear/train.py (hunk 2 above)
Comment (weiliw-amz) [reply pecos/distributed/xmc/xlinear/train.py:174]: Yea we should do that, whether to include it into this PR or leave for future improvement is left to debate.

File: pecos/distributed/comm/dummy_comm.py (hunk 1 above)
Comment (weiliw-amz) [reply pecos/distributed/comm/dummy_comm.py]: As discussed, moved to `distributed.diagnostic_util`

File: pecos/distributed/xmc/xlinear/train.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        help="the minimum number of sub-trees in training step, should be more than number of distributed machines.",
+    )
+    parser.add_argument(
+        "--nr-splits",
+        type=int,
+        default=2,
+        metavar="INT",
+        help="number of splits used to construct hierarchy (a power of 2 is recommended)",
+    )
+    parser.add_argument(
```
Comment (jiong-zhang) [inline pecos/distributed/xmc/xlinear/train.py:70]: re-rank the argument like `pecos.xmc.xlinear.train`, i.e. required args first.

File: pecos/distributed/xmc/xlinear/train.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        metavar="INT",
+        help="number of threads to use (default -1 to denote all the CPUs)",
+    )
+    parser.add_argument(
+        "-ms",
+        "--main-speed",
+        type=float,
+        default=0.3,
+        metavar="FLOAT",
+        help="main node cpu speed",
```
Comment (jiong-zhang) [inline pecos/distributed/xmc/xlinear/train.py]: We need more explanation of this arg.

File: pecos/distributed/xmc/xlinear/train.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    parser.add_argument(
+        "--verbose-level",
+        type=int,
+        choices=logging_util.log_levels.keys(),
+        default=2,
+        metavar="INT",
+        help=f"the verbose level, {', '.join([str(k) + ' for ' + logging.getLevelName(v) for k, v in logging_util.log_levels.items()])}. Default 2",
+    )
+
+    return parser
```
Comment (jiong-zhang) [inline pecos/distributed/xmc/xlinear/train.py:163]: Arguments to add in the future to be consistent with `xlinear`
* Simple:
    * *[?]--code-path*:  support user given clusters and skip dist-clustering
    * *[?]--Cp, --Cn, --solver-type*: parameters
    * *[?]--label-feat-path:* user given label features


* Medium:
    * *Cost Sensitive Learning*: 
        * *--rel-path, --rel-norm, --rel-mode*
* Hard:
    * User supplied negatives:
        * *--usn-label-path, --usn-match-path*

**Review by jiong-zhang — APPROVED** [review]

**Pull request merged**